### ISSUE TITLE
-Wsign-compare false positive with comparison against literal -1

### ISSUE THREAD

**Issue by alejandro-colomar** [llvm/llvm-project#128746]
In C, comparing an unsigned integer (of rank no less than that of `int`) to a literal `-1` is a common idiom in error-handling code, and is harmless.  The -1 is first sign-extended to the width of the unsigned integer, and then is converted into an unsigned value, which is The Right Thing to do.

However, `-Wsign-compare` (in `-Wextra`) diagnoses about it:

```c
alx@debian:~/tmp$ cat f.c
int
main(void)
{
	unsigned long i = 3;

	if (i == -1)
		return 1;

	return 0;
}
alx@debian:~/tmp$ clang -Wsign-compare f.c 
f.c:6:8: warning: comparison of integers of different signs: 'unsigned long' and 'int' [-Wsign-compare]
    6 |         if (i == -1)
      |             ~ ^  ~~
1 warning generated.
```

A user wanting to turn off the diagnostic, may choose to add a cast somewhere in that comparison, and chances are decent that it will put it in the non-literal part (I've seen code like that).  And the problem is that such code has a bug, which cannot even be diagnosed (because the cast turns off the diagnostic).

Or maybe it adds a cast in the right place, but the type of the cast is erroneous, which would similarly result in a silent bug.

Another approach that a user might take, to prevent both the diagnostic and a cast, would be to use a suffix in the literal, such as `if (i == -1u)`.  This will similarly result in a silent (non-diagnosable) bug.  And since we don't have a way to create a suffix for a `time_t`, the programmer has to do a good guess (or write a lot of macros that might get it correctly.

All workarounds result in worse code than a literal `-1`, both in terms of safety, and in terms of readability.

Let's just not warn when comparing an unsigned integer against literal `-1` (with no suffixes), as long as the unsigned integer type has rank no less than `int` (so that the promotions result in an unsigned type, and the -1 is sign-extended).

Here's a bug I opened against GCC: <https://gcc.gnu.org/bugzilla/show_bug.cgi?id=119011>

Cc: @jwakely , @AaronBallman , @ThePhD
